src/metabase/native_query_snippets/cycle_detection.clj

```diff
@@ -0,0 +1,68 @@
+(ns metabase.native-query-snippets.cycle-detection
+  "Cycle detection for native query snippets.
+   Handles pure snippet→snippet cycles directly, delegates to lib/query for card-involved cycles."
+  (:require
+   [metabase.lib-be.metadata.jvm :as lib.metadata.jvm]
+   [metabase.lib.core :as lib]
+   [metabase.queries.core :as queries]
+   [toucan2.core :as t2]))
+
+(defn- ->snippet-dependencies
+  "Returns sequence of [:card card-id] and [:snippet snippet-name]."
+  [template-tags]
+  (for [[_ tag] template-tags
+        :let [tag-type (:type tag)]
+        :when (#{:card :snippet} tag-type)]
+    (case tag-type
+      :card [tag-type (:card-id tag)]
+      :snippet [tag-type (:snippet-name tag)])))
+
+(defn- names->ids-and-tags
+  "Accepts a sequence of snippet-names. Returns a sequence of [snippet-id snippe-template-tags]."
+  [snippet-names]
+  (when (seq snippet-names)
+    (for [snippet (t2/select :model/NativeQuerySnippet :name [:in snippet-names])]
```

Comment:
IMO you should fetch this from a metadata provider so we can easily mock it (`bulk-metadata` lets you efficiently fetch a lot of stuff at once).

If you were using Toucan don't forget you can do 

```clj
(t2/select-fn-vec (juxt :id :template_tags) :model/NativeQuerySnippet :name [:in snippet-names])
```

so you're not fetching the entire row. (See my friendly reminder https://metaboat.slack.com/archives/CKZEMT1MJ/p1755287727177819)